### ISSUE TITLE
Linker errors when compiling test with nmake

### ISSUE THREAD

**Issue by ash-j-f** [jtv/libpqxx#216]
I see the following linker errors when I try "nmake /f win32\vc-test.mak ALL"

```
test00.obj : error LNK2019: unresolved external symbol "public: static double __cdecl pqxx::internal::builtin_traits<double>::from_string(class std::basic_string_view<char,struct std::char_traits<char> >)" (?from_string@?$builtin_traits@N@internal@pqxx@@SANV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z) referenced in function "void __cdecl pqxx::from_string<double>(class std::basic_string_view<char,struct std::char_traits<char> >,double &)" (??$from_string@N@pqxx@@YAXV?$basic_string_view@DU?$char_traits@D@std@@@std@@AEAN@Z)
test46.obj : error LNK2001: unresolved external symbol "public: static double __cdecl pqxx::internal::builtin_traits<double>::from_string(class std::basic_string_view<char,struct std::char_traits<char> >)" (?from_string@?$builtin_traits@N@internal@pqxx@@SANV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
test74.obj : error LNK2001: unresolved external symbol "public: static double __cdecl pqxx::internal::builtin_traits<double>::from_string(class std::basic_string_view<char,struct std::char_traits<char> >)" (?from_string@?$builtin_traits@N@internal@pqxx@@SANV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
test00.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<double>(double)" (??$to_string_float@N@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@N@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::to_string<double>(double const &)" (??$to_string@N@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBN@Z)
test74.obj : error LNK2001: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<double>(double)" (??$to_string_float@N@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@N@Z)
test46.obj : error LNK2019: unresolved external symbol "public: static float __cdecl pqxx::internal::builtin_traits<float>::from_string(class std::basic_string_view<char,struct std::char_traits<char> >)" (?from_string@?$builtin_traits@M@internal@pqxx@@SAMV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z) referenced in function "void __cdecl pqxx::from_string<float>(class std::basic_string_view<char,struct std::char_traits<char> >,float &)" (??$from_string@M@pqxx@@YAXV?$basic_string_view@DU?$char_traits@D@std@@@std@@AEAM@Z)
test74.obj : error LNK2001: unresolved external symbol "public: static float __cdecl pqxx::internal::builtin_traits<float>::from_string(class std::basic_string_view<char,struct std::char_traits<char> >)" (?from_string@?$builtin_traits@M@internal@pqxx@@SAMV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
test46.obj : error LNK2019: unresolved external symbol "public: static long double __cdecl pqxx::internal::builtin_traits<long double>::from_string(class std::basic_string_view<char,struct std::char_traits<char> >)" (?from_string@?$builtin_traits@O@internal@pqxx@@SAOV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z) referenced in function "void __cdecl pqxx::from_string<long double>(class std::basic_string_view<char,struct std::char_traits<char> >,long double &)" (??$from_string@O@pqxx@@YAXV?$basic_string_view@DU?$char_traits@D@std@@@std@@AEAO@Z)
test74.obj : error LNK2001: unresolved external symbol "public: static long double __cdecl pqxx::internal::builtin_traits<long double>::from_string(class std::basic_string_view<char,struct std::char_traits<char> >)" (?from_string@?$builtin_traits@O@internal@pqxx@@SAOV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
test74.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<float>(float)" (??$to_string_float@M@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@M@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::to_string<float>(float const &)" (??$to_string@M@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBM@Z)
test74.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<long double>(long double)" (??$to_string_float@O@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@O@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::to_string<long double>(long double const &)" (??$to_string@O@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBO@Z)
TestStaticRelease\\runner.exe : fatal error LNK1120: 6 unresolved externals
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\link.exe"' : return code '0x460'
Stop.
```

Original discussion for this issue was started here https://github.com/jtv/libpqxx/issues/213#issuecomment-526519587

@jtv made some updates here in eaaa3cc to try to resolve the issue (added a missing PQXX_LIBEXPORT) but I noted that this did not resolve the issue.

**Comment by ash-j-f** [comment]
It may be useful to know that if I comment out the following in config-internal-compiler.h as shown, I still get the same error.
```
//#define PQXX_HAVE_CHARCONV_INT 1
//#define PQXX_HAVE_CHARCONV_FLOAT 1
```

**Comment by jtv** [comment]
Found a good reason why you might get these link errors when `PQXX_HAVE_CHARCONV_FLOAT` is defined.  Still no idea why you'd get them when it's not defined though.  :-/

Does my latest commit at least help the case where it's defined?

**Comment by ash-j-f** [comment]
That's an improvement! Now down to 3 unresolved externals instead of 6.
```
nmake /f win32\vc-test.mak ALL
...
test00.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<double>(double)" (??$to_string_float@N@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@N@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::to_string<double>(double const &)" (??$to_string@N@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBN@Z)
test74.obj : error LNK2001: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<double>(double)" (??$to_string_float@N@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@N@Z)
test74.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<float>(float)" (??$to_string_float@M@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@M@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::to_string<float>(float const &)" (??$to_string@M@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBM@Z)
test74.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::internal::to_string_float<long double>(long double)" (??$to_string_float@O@internal@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@O@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl pqxx::to_string<long double>(long double const &)" (??$to_string@O@pqxx@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBO@Z)
TestStaticRelease\\runner.exe : fatal error LNK1120: 3 unresolved externals
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\link.exe"' : return code '0x460'
Stop.
```

**Comment by ash-j-f** [comment]
I tried commenting out the PQXX_HAVE_CHARCONV_INT and PQXX_HAVE_CHARCONV_FLOAT in config-internal-compiler.h again and this time the test suite compiles and runs with no errors. So maybe last time I screwed something up (maybe I forgot to run a clean before the new build?), or recent changes stopped that strange bug.

**Comment by jtv** [comment]
Getting there!  It led me to a spot where I had a `PQXX_LIBEXPORT` that shouldn't have been there.  Any further improvement from my latest commit?

**Comment by ash-j-f** [comment]
I still get the same errors, but I think I found a reason why. to_string_float is never declared or defined if PQXX_HAVE_CHARCONV_FLOAT is set.

Both instances I could find declaring to_string_float are inside the blocks ```#if !defined(PQXX_HAVE_CHARCONV_FLOAT)```. So they will never be used by the compiler when PQXX_HAVE_CHARCONV_FLOAT is set to 1.

The places I can see affected are:
conversions.hxx line 186
strconv.cxx line 342

Simply allowing those definitions to be used causes an error of dumb_stringstream being an unknown type, so I'd guess the fix is more complicated than simply allowing those definitions to be used. :)

**Comment by jtv** [comment]
Thanks.  This stuff makes me a little snowblind sometimes!

Hoping to fix it tonight.

**Comment by jtv** [comment]
I went down the rabbit hole on that one, trying to reform the whole `to_string` constellation.  Then I started again from scratch, and found that _the headers_ check for `PQXX_HAVE_CHARCONV_FLOAT`.  But in my zeal to minimise namespace pollution, I never made that information available to the headers!  By design, I hide that implementation detail completely inside the library binary.  But not completely, it turns out.

Does my latest commit fix it?

**Comment by ash-j-f** [comment]
Works like a charm! All the tests complete with no errors, including DLL and DEBUG variations. Thanks for all your hard work. :D Any plans for another 6.x.x release? Or will the next one be a 7.0? :D

**Comment by jtv** [comment]
Thank _you_ for your persistent and patient debugging and reporting.

The only plan right now is to go to 7.0, but it's a big project.  But if there's a really urgent bug fix for 6.0, I can still release updated versions.

**Comment by jtv** [comment]
Update: I've got the CMake build working on AppVeyor.  And since this shows that the CMake build can work on Windows, I have _removed the custom Visual C++ build files._

The one big remaining problem with the CMake build right now is that it only builds a shared library.  We'll fix that.
